{"nodes":[{"pos":[12,69],"content":"Mixing C (Structured) and C++ Exceptions | Microsoft Docs","needQuote":false,"needEscape":true,"nodes":[{"content":"Mixing C (Structured) and C++ Exceptions | Microsoft Docs","pos":[0,57]}]},{"content":"Mixing C (Structured) and C++ Exceptions","pos":[793,833]},{"content":"If you want to write more portable code, using structured exception handling in a C++ program is not recommended.","pos":[834,947]},{"content":"However, you may sometimes want to compile with <bpt id=\"p1\">**</bpt>/EHa<ept id=\"p1\">**</ept> and mix structured exceptions and C++ source code, and need some facility for handling both kinds of exceptions.","pos":[948,1117],"source":" However, you may sometimes want to compile with **/EHa** and mix structured exceptions and C++ source code, and need some facility for handling both kinds of exceptions."},{"content":"Because a structured exception handler has no concept of objects or typed exceptions, it cannot handle exceptions thrown by C++ code; however, C++ <bpt id=\"p1\">**</bpt>catch<ept id=\"p1\">**</ept> handlers can handle structured exceptions.","pos":[1118,1317],"source":" Because a structured exception handler has no concept of objects or typed exceptions, it cannot handle exceptions thrown by C++ code; however, C++ **catch** handlers can handle structured exceptions."},{"content":"As such, C++ exception handling syntax (<bpt id=\"p1\">**</bpt>try<ept id=\"p1\">**</ept>, <ph id=\"ph1\">`throw`</ph>, <bpt id=\"p2\">**</bpt>catch<ept id=\"p2\">**</ept>) is not accepted by the C compiler, but structured exception handling syntax (<ph id=\"ph2\">`__try`</ph>, <ph id=\"ph3\">`__except`</ph>, <ph id=\"ph4\">`__finally`</ph>) is supported by the C++ compiler.","pos":[1318,1531],"source":" As such, C++ exception handling syntax (**try**, `throw`, **catch**) is not accepted by the C compiler, but structured exception handling syntax (`__try`, `__except`, `__finally`) is supported by the C++ compiler."},{"pos":[1538,1684],"content":"See <bpt id=\"p1\">[</bpt>_set_se_translator<ept id=\"p1\">](../c-runtime-library/reference/set-se-translator.md)</ept> for information on handling structured exceptions as C++ exceptions.","source":"See [_set_se_translator](../c-runtime-library/reference/set-se-translator.md) for information on handling structured exceptions as C++ exceptions."},{"content":"If you mix structured and C++ exceptions, note the following:","pos":[1691,1752]},{"content":"C++ exceptions and structured exceptions cannot be mixed within the same function.","pos":[1762,1844]},{"pos":[1854,1968],"content":"Termination handlers (<ph id=\"ph1\">`__finally`</ph> blocks) are always executed, even during unwinding after an exception is thrown.","source":"Termination handlers (`__finally` blocks) are always executed, even during unwinding after an exception is thrown."},{"pos":[1978,2186],"content":"C++ exception handling can catch and preserve unwind semantics in all modules compiled with the <bpt id=\"p1\">[</bpt>/EH<ept id=\"p1\">](../build/reference/eh-exception-handling-model.md)</ept> compiler option (this option enables unwind semantics).","source":"C++ exception handling can catch and preserve unwind semantics in all modules compiled with the [/EH](../build/reference/eh-exception-handling-model.md) compiler option (this option enables unwind semantics)."},{"content":"There may be some situations in which destructor functions are not called for all objects.","pos":[2196,2286]},{"content":"For example, if a structured exception occurs while attempting to make a function call through an uninitialized function pointer, and that function takes as parameters objects that were constructed before the call, those objects will not have their destructors called during stack unwind.","pos":[2287,2575]},{"content":"What do you want to know more about?","pos":[2584,2620]},{"content":"Using setjmp or longjmp in C++ programs","pos":[2631,2670]},{"content":"Differences between SEH and C++ EH","pos":[2714,2748]},{"content":"See Also","pos":[2800,2808]},{"content":"C++ Exception Handling","pos":[2813,2835]}],"content":"---\ntitle: \"Mixing C (Structured) and C++ Exceptions | Microsoft Docs\"\nms.custom: \"\"\nms.date: \"11/04/2016\"\nms.reviewer: \"\"\nms.suite: \"\"\nms.technology: \n  - \"devlang-cpp\"\nms.tgt_pltfrm: \"\"\nms.topic: \"language-reference\"\ndev_langs: \n  - \"C++\"\nhelpviewer_keywords: \n  - \"exceptions, mixed C and C++\"\n  - \"C++ exception handling, mixed-language\"\n  - \"structured exception handling, mixed C and C++\"\n  - \"catch keyword [C++], mixed\"\n  - \"try-catch keyword [C++], mixed-language\"\nms.assetid: a149154e-36dd-4d1a-980b-efde2a563a56\ncaps.latest.revision: 7\nauthor: \"mikeblome\"\nms.author: \"mblome\"\nmanager: \"ghogen\"\ntranslation.priority.ht: \n  - \"cs-cz\"\n  - \"de-de\"\n  - \"es-es\"\n  - \"fr-fr\"\n  - \"it-it\"\n  - \"ja-jp\"\n  - \"ko-kr\"\n  - \"pl-pl\"\n  - \"pt-br\"\n  - \"ru-ru\"\n  - \"tr-tr\"\n  - \"zh-cn\"\n  - \"zh-tw\"\n---\n# Mixing C (Structured) and C++ Exceptions\nIf you want to write more portable code, using structured exception handling in a C++ program is not recommended. However, you may sometimes want to compile with **/EHa** and mix structured exceptions and C++ source code, and need some facility for handling both kinds of exceptions. Because a structured exception handler has no concept of objects or typed exceptions, it cannot handle exceptions thrown by C++ code; however, C++ **catch** handlers can handle structured exceptions. As such, C++ exception handling syntax (**try**, `throw`, **catch**) is not accepted by the C compiler, but structured exception handling syntax (`__try`, `__except`, `__finally`) is supported by the C++ compiler.  \n  \n See [_set_se_translator](../c-runtime-library/reference/set-se-translator.md) for information on handling structured exceptions as C++ exceptions.  \n  \n If you mix structured and C++ exceptions, note the following:  \n  \n1.  C++ exceptions and structured exceptions cannot be mixed within the same function.  \n  \n2.  Termination handlers (`__finally` blocks) are always executed, even during unwinding after an exception is thrown.  \n  \n3.  C++ exception handling can catch and preserve unwind semantics in all modules compiled with the [/EH](../build/reference/eh-exception-handling-model.md) compiler option (this option enables unwind semantics).  \n  \n4.  There may be some situations in which destructor functions are not called for all objects. For example, if a structured exception occurs while attempting to make a function call through an uninitialized function pointer, and that function takes as parameters objects that were constructed before the call, those objects will not have their destructors called during stack unwind.  \n  \n## What do you want to know more about?  \n  \n-   [Using setjmp or longjmp in C++ programs](../cpp/using-setjmp-longjmp.md)  \n  \n-   [Differences between SEH and C++ EH](../cpp/exception-handling-differences.md)  \n  \n## See Also  \n [C++ Exception Handling](../cpp/cpp-exception-handling.md)"}